File: src/org/elixir_lang/reference/resolver/atom/resolvable/Exact.java
package org.elixir_lang.reference.resolver.atom.resolvable;

import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElementResolveResult;
import com.intellij.psi.ResolveResult;
import org.elixir_lang.reference.resolver.atom.Resolvable;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import static org.elixir_lang.Reference.forEachNavigationElement;

public class Exact extends Resolvable {
    @NotNull
    private final String name;

    public Exact(@NotNull String name) {
        this.name = name;
    }

    @Override
    public ResolveResult[] resolve(@NotNull Project project) {
        List<ResolveResult> resolveResultList = new ArrayList<>();

        forEachNavigationElement(
                project,
                name,
                navigationElement -> {
                    resolveResultList.add(new PsiElementResolveResult(navigationElement));

                    return true;
                }
        );

        return resolveResultList.toArray(new ResolveResult[resolveResultList.size()]);
    }
}

File: src/org/elixir_lang/reference/resolver/atom/resolvable/Exact.kt
package org.elixir_lang.reference.resolver.atom.resolvable

import com.intellij.openapi.project.DumbService
import com.intellij.psi.PsiElementResolveResult
import com.intellij.psi.ResolveResult
import org.elixir_lang.psi.ElixirAtom
import org.elixir_lang.psi.NamedElement
import org.elixir_lang.psi.stub.index.AllName
import org.elixir_lang.reference.resolver.atom.Resolvable
import org.elixir_lang.reference.resolver.narrowedScope
import com.intellij.psi.stubs.StubIndex

class Exact(private val name: String) : Resolvable() {
    override fun resolve(element: ElixirAtom): Array<ResolveResult> {
        val resolveResultList = mutableListOf<ResolveResult>()
        val project = element.project

        if (!DumbService.isDumb(project)) {
            val scope = narrowedScope(element, project)

            StubIndex.getInstance().processElements(
                AllName.KEY, name, project, scope, NamedElement::class.java
            ) { namedElement ->
                resolveResultList.add(PsiElementResolveResult(namedElement))

                true
            }
        }

        return resolveResultList.toTypedArray()
    }
}
